File: app/services/codex/exec_runner.rb
require "open3"
require "json"
require "tmpdir"
require "timeout"

class Codex::ExecRunner
  DEFAULT_SANDBOX_MODE = "danger-full-access"
  ExecutionError = Class.new(StandardError)
  ExecutionTimeoutError = Class.new(ExecutionError)

  def initialize(timeout_seconds: AppConfig.fetch.codex_exec_timeout_seconds)
    @timeout_seconds = timeout_seconds
  end

  def run(prompt:, image_file_paths: [], output_schema: nil)
    Dir.mktmpdir("telegram-codex-") do |dir|
      output_path = File.join(dir, "reply.txt")
      schema_path = build_schema_file(dir, output_schema)
      command = build_command(output_path, image_file_paths, schema_path)

      _stdout, stderr, status = run_command(command, prompt)
      raise ExecutionError, "codex exec failed: #{stderr.strip.presence || 'unknown error'}" unless status.success?

      reply_text = File.read(output_path).strip
      raise ExecutionError, "codex exec returned an empty reply" if reply_text.empty?

      reply_text
    end
  end

  private

  attr_reader :timeout_seconds

  def run_command(command, prompt)
    stdout = +""
    stderr = +""
    status = nil

    Open3.popen3(*command, chdir: Rails.root.to_s) do |stdin, stdout_stream, stderr_stream, wait_thread|
      stdin.write(prompt)
      stdin.close

      stdout_reader = Thread.new { stdout_stream.read.to_s }
      stderr_reader = Thread.new { stderr_stream.read.to_s }

      begin
        status = Timeout.timeout(timeout_seconds) { wait_thread.value }
      rescue Timeout::Error
        terminate_process(wait_thread.pid)
        raise ExecutionTimeoutError, "codex exec timed out after #{timeout_seconds} seconds"
      ensure
        stdout = stdout_reader.value
        stderr = stderr_reader.value
      end
    end

    [stdout, stderr, status]
  rescue Timeout::Error
    raise ExecutionTimeoutError, "codex exec timed out after #{timeout_seconds} seconds"
  end

  def terminate_process(pid)
    Process.kill("TERM", pid)
    Timeout.timeout(2) { Process.wait(pid) }
  rescue Errno::ESRCH, Errno::ECHILD
    nil
  rescue Timeout::Error
    Process.kill("KILL", pid)
  rescue Errno::ESRCH
    nil
  end

  def build_command(output_path, image_file_paths, schema_path)
    sandbox_mode = ENV.fetch("CODEX_SANDBOX_MODE", DEFAULT_SANDBOX_MODE)
    command = [
      "codex",
      "exec",
      "--skip-git-repo-check",
      "--sandbox", sandbox_mode,
      "--color", "never",
      "--output-last-message", output_path
    ]
    command += ["--output-schema", schema_path] if schema_path.present?
    Array(image_file_paths).each do |image_file_path|
      command += ["--image", image_file_path] if image_file_path.present?
    end
    command << "-"
    command
  end

  def build_schema_file(dir, output_schema)
    return nil if output_schema.blank?

    schema_path = File.join(dir, "reply-schema.json")
    File.write(schema_path, JSON.generate(output_schema))
    schema_path
  end
end

File: src/main/java/com/telegram/codex/codex/ExecRunner.java
package com.telegram.codex.codex;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.telegram.codex.config.AppProperties;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

@Component
public class ExecRunner {

    public static class ExecutionError extends RuntimeException {
        public ExecutionError(String message) {
            super(message);
        }

        public ExecutionError(String message, Throwable cause) {
            super(message, cause);
        }
    }

    public static class ExecutionTimeoutError extends ExecutionError {
        public ExecutionTimeoutError(String message) {
            super(message);
        }
    }

    private static final String DEFAULT_SANDBOX_MODE = "danger-full-access";

    private final AppProperties properties;
    private final ObjectMapper objectMapper;

    public ExecRunner(AppProperties properties, ObjectMapper objectMapper) {
        this.properties = properties;
        this.objectMapper = objectMapper;
    }

    public String run(String prompt, List<Path> imageFilePaths, Map<String, Object> outputSchema) {
        try {
            Path tempDir = Files.createTempDirectory("telegram-codex-");
            Path outputPath = tempDir.resolve("reply.txt");
            Path schemaPath = outputSchema == null ? null : tempDir.resolve("reply-schema.json");
            if (schemaPath != null) {
                Files.writeString(schemaPath, objectMapper.writeValueAsString(outputSchema), StandardCharsets.UTF_8);
            }

            List<String> command = new ArrayList<>();
            command.add("codex");
            command.add("exec");
            command.add("--skip-git-repo-check");
            command.add("--sandbox");
            command.add(System.getenv().getOrDefault("CODEX_SANDBOX_MODE", DEFAULT_SANDBOX_MODE));
            command.add("--color");
            command.add("never");
            command.add("--output-last-message");
            command.add(outputPath.toString());
            if (schemaPath != null) {
                command.add("--output-schema");
                command.add(schemaPath.toString());
            }
            for (Path imageFilePath : imageFilePaths) {
                command.add("--image");
                command.add(imageFilePath.toString());
            }
            command.add("-");

            Process process = new ProcessBuilder(command)
                .directory(Path.of(".").toFile())
                .start();
            process.getOutputStream().write(prompt.getBytes(StandardCharsets.UTF_8));
            process.getOutputStream().close();

            boolean exited = process.waitFor(properties.getCodexExecTimeoutSeconds(), TimeUnit.SECONDS);
            String stderr = new BufferedReader(new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8))
                .lines()
                .reduce("", (left, right) -> left + right + "\n");
            if (!exited) {
                process.destroyForcibly();
                throw new ExecutionTimeoutError("codex exec timed out after " + properties.getCodexExecTimeoutSeconds() + " seconds");
            }
            if (process.exitValue() != 0) {
                throw new ExecutionError("codex exec failed: " + (stderr.isBlank() ? "unknown error" : stderr.trim()));
            }
            String replyText = Files.readString(outputPath, StandardCharsets.UTF_8).trim();
            if (replyText.isEmpty()) {
                throw new ExecutionError("codex exec returned an empty reply");
            }
            return replyText;
        } catch (ExecutionError error) {
            throw error;
        } catch (Exception error) {
            throw new ExecutionError("Failed to run codex exec", error);
        }
    }
}
